apps/rush-lib/src/pluginFramework/PluginManager.ts

```diff
@@ -0,0 +1,206 @@
+  }
+
+  public async _preparePluginAutoinstallersAsync(pluginLoaders: RemotePluginLoader[]): Promise<void> {
+    for (const { autoinstaller } of pluginLoaders) {
+      if (!this._installedAutoinstallerNames.has(autoinstaller.name)) {
+        await autoinstaller.prepareAsync();
+        this._installedAutoinstallerNames.add(autoinstaller.name);
+      }
+    }
+  }
+
+  public async tryInitializeUnassociatedPluginsAsync(): Promise<void> {
+    try {
+      const remotePluginLoaders: RemotePluginLoader[] = this._getUnassociatedPluginLoaders(
+        this._remotePluginLoaders
+      );
+      await this._preparePluginAutoinstallersAsync(remotePluginLoaders);
+      const defaultPluginLoaders: DefaultPluginLoader[] = this._getUnassociatedPluginLoaders(
+        this._defaultPluginLoaders
+      );
+      await this._initializePluginsAsync([...defaultPluginLoaders, ...remotePluginLoaders]);
+    } catch (e) {
+      this._error = e as Error;
+    }
+  }
+
+  public async tryInitializeAssociatedCommandPluginsAsync(commandName: string): Promise<void> {
+    try {
+      const remotePluginLoaders: RemotePluginLoader[] = this._getPluginLoadersForCommand(
+        commandName,
+        this._remotePluginLoaders
+      );
+      await this._preparePluginAutoinstallersAsync(remotePluginLoaders);
+      const defaultPluginLoaders: DefaultPluginLoader[] = this._getPluginLoadersForCommand(
+        commandName,
+        this._defaultPluginLoaders
+      );
+      await this._initializePluginsAsync([...defaultPluginLoaders, ...remotePluginLoaders]);
+    } catch (e) {
+      this._error = e as Error;
+    }
+  }
+
+  public tryGetCustomCommandLineConfigurationInfos(): ICustomCommandLineConfigurationInfo[] {
+    const commandLineConfigurationInfos: ICustomCommandLineConfigurationInfo[] = [];
+    for (const pluginLoader of this._remotePluginLoaders) {
+      const commandLineConfiguration: CommandLineConfiguration | undefined =
+        pluginLoader.getCommandLineConfiguration();
+      if (commandLineConfiguration) {
+        commandLineConfigurationInfos.push({
+          commandLineConfiguration,
+          pluginLoader
+        });
+      }
+    }
+    return commandLineConfigurationInfos;
+  }
+
+  private async _initializePluginsAsync(pluginLoaders: PluginLoaderBase[]): Promise<void> {
+    const pluginInfos: { plugin: IRushPlugin; pluginName: string }[] = [];
+    for (const pluginLoader of pluginLoaders) {
+      const pluginName: string = pluginLoader.pluginName;
+      if (this._loadedPluginNames.has(pluginName)) {
+        throw new Error(
+          `Plugin "${pluginName}" has already been loaded. Please check your rush-plugins.json.`
+        );
+      }
+      const plugin: IRushPlugin | undefined = pluginLoader.load();
+      if (plugin) {
+        pluginInfos.push({
+          pluginName,
+          plugin
+        });
+        this._loadedPluginNames.add(pluginName);
+      }
+    }
+    for (const { plugin, pluginName } of pluginInfos) {
```

Comment:
Is it important that we fully check for already loaded plugins before doing this? That seems like we're optimizing for the failure case when we could instead just apply the plugins as we go.